File: java/src/main/java/com/leetcode/crackcodeinterview/ZeroOneZeroOne.java
package com.leetcode.crackcodeinterview;

/**
 * @author: John
 * @date: 2020-09-14 15:30
 * @description: 0101
 **/
public class ZeroOneZeroOne {
//    public boolean isUnique(String astr) {
//        for (int i = 0; i < astr.length(); i++) {
//            if (astr.lastIndexOf(astr.charAt(i)) != i) {
//                return false;
//            }
//        }
//        return true;
//    }

    public boolean isUnique(String astr) {
       long mark = 0;
       for (int i = 0; i < astr.length(); i++) {
           int num = 1 << (astr.charAt(i) - 'a');
           if ((mark & num) != 0) {
               return false;
           } else {
               mark |= num;
           }
       }
       return true;
    }


}

File: leetcodecpp/leetcode/crackcodeinterview/ZeroOneZeroOne.cpp
//
// Created by John on 2020/9/14.
//
#include <iostream>
#include "string"

using namespace std;

class Solution {
public:
    bool isUnique(string astr) {
        for (int i = 0; i < astr.size(); i++) {
            if (astr.find(astr[i]) != astr.rfind(astr[i])) {
//                printf(astr.find(astr[i]), "%s");
//                printf(astr.find(astr[i]), "%s");
                cout << astr.find(astr[i])  << "\n";
                cout << astr.rfind(astr[i]) << "\n";
                return false;
            }
        }
        return true;
    }
};
//int main() {
//    Solution * solution = new Solution();
//    solution->isUnique("leetcode");
//}

